cpp/arcticdb/entity/metrics.hpp

```diff
@@ -114,26 +160,61 @@ class PrometheusInstance {
     private:
 
         struct HistogramInfo {
-            prometheus::Family<prometheus::Histogram>* histogram;
-            prometheus::Histogram::BucketBoundaries buckets_list;
+            HistogramInfo(prometheus::Family<prometheus::Histogram>* histogram,
+                          prometheus::Histogram::BucketBoundaries buckets_list) : histogram_(histogram),
+                          buckets_list_(std::move(buckets_list)) {
+
+            }
+
+            prometheus::Family<prometheus::Histogram>* histogram_;
+            prometheus::Histogram::BucketBoundaries buckets_list_;
+        };
+
+        template<typename T>
+        using MetricFamilyMap = std::unordered_map<std::string, prometheus::Family<T>*>;
+
+        struct MetricKey {
+            std::string name;
+            Labels labels;
+
+            bool operator==(const MetricKey &) const = default;
+        };
+
+        struct MetricKeyHash {
+            std::size_t operator()(const MetricKey& key) const noexcept {
+                auto hash = std::hash<std::string>()(key.name);
+                for (const auto& [name, value] : key.labels) {
+                    hash = folly::hash::commutative_hash_combine(hash, name, value);
+                }
+                return hash;
+            }
         };
 
+        template<typename T>
+        using MetricMap = std::unordered_map<MetricKey, T*, MetricKeyHash>;
+
         static std::string getHostName();
 
         std::shared_ptr<prometheus::Registry> registry_;
         std::shared_ptr<prometheus::Exposer> exposer_;
-        // cannot use ref type, so use pointer
-        std::unordered_map<std::string, prometheus::Family<prometheus::Counter>*> map_counter_;
-        std::unordered_map<std::string, prometheus::Family<prometheus::Gauge>*> map_gauge_;
+
+        MetricFamilyMap<prometheus::Counter> map_counter_;
+        MetricMap<prometheus::Counter> all_counters_;
+
+        MetricFamilyMap<prometheus::Gauge> map_gauge_;
+        MetricMap<prometheus::Gauge> all_gauges_;
+
         std::unordered_map<std::string, HistogramInfo> map_histogram_;
-        prometheus::Histogram::BucketBoundaries buckets_list_;
+        MetricMap<prometheus::Histogram> all_histograms_;
+
+        MetricFamilyMap<prometheus::Summary> map_summary_;
+        MetricMap<prometheus::Summary> all_summaries_;
 
-        std::unordered_map<std::string, prometheus::Family<prometheus::Summary>*> map_summary_;
-        // push gateway
         std::string mongo_instance_;
         std::shared_ptr<prometheus::Gateway> gateway_;
         bool configured_;
 
+        std::mutex metrics_mutex_;
```

Comment:
The threading model used to be:

- "Main" thread - creates Prometheus instance, registers metrics. This populates `map_counter_` etc
- The Prometheus SDK is thread-safe. Registering the metric takes a lock, which establishes a memory order.
- Other threads are then free to set the metric values, eg calling `incrementCounter`. I think there was a pre-existing synchronization issue though - there is no guarantee the other threads will see the correct state of `map_counter_`.
- The methods that set the metric values did not modify any state except that of Prometheus SDK objects, which are threadsafe.

Since now the methods setting the metric values modifies state in the `PrometheusInstance`, we need additional synchronisation.